---
title: "DESeq2_analysis_pipeline"
author: "Alisa Fedorenko"
date: "2024-05-20"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r message=FALSE, warning=FALSE}
require(AnnotationHub)
require(ensembldb)
require(dplyr)
require(stringr)
require(tximport)
require(ggplot2)
require(tibble)
require(DESeq2)
require(pheatmap)
require(EnhancedVolcano)
require(clusterProfiler)
require(org.Hs.eg.db)
require(forcats)
require(pathview)
require(tidyverse)
require(readxl)
require(EnsDb.Hsapiens.v86)
require(ggpubr)
require(ggrepel)
```

```{r}
#setwd("./path/to/the/folder")
```

Provide all the data of salmon output in 'salmon_output' folder. To make an order in the directory create 'data' folder for all output tables and 'plots' folder for all output plots.

```{r warning=TRUE}
dir.create("./data")
dir.create("./plots")
```

## STEP1. Annotation.

```{r}
# Connect to AnnotationHub:
ah <- AnnotationHub()

# Return the Ensembl EnsDb information for Homo Sapiens:
human_ens <- query(ah, c("Homo Sapiens", "EnsDb"))
human_ens

# Extract the latest Ensembl release using the AnnotationHub ID to subset the object
human_ens <- human_ens[["AH109606"]]

# Now we can use functions from ensembldb package to extract the information at the gene, transcript, or exon levels. We are interested in the gene-level annotations, so we can extract that information as follows:

genedb <- genes(human_ens, return.type = "data.frame")

# Look at the first 3 rows:
head(genedb, 3)

#Select columns “gene_id” and “symbol”:
genedb2 <- dplyr::select(genedb, gene_id, symbol)

# Since we need to convert transcript IDs to gene IDs, we also should extract the information at the transcript level.
txdb <- transcripts(human_ens, return.type = "data.frame")
head(txdb)

# Select columns “tx_id” and “gene_id” from txdb and save it to txdb2 object as mentioned above.
# Check the first 5 rows of the table.
txdb2 <- dplyr::select(txdb, tx_id, gene_id)
head(txdb2, 5)

# Use dplyr R package to join transcripts, genes ID and symbols into one table: Transcript ID must be first column!
tx2gene <- right_join(txdb2, genedb2, by='gene_id')
head(tx2gene)

# The final table must have 3 columns. Save the annotation table:
write.table(tx2gene, "./data/annotation_human.csv", sep=';', quote=F, row.names=F)
```

## STEP2. Tximport.

```{r}
# List all directories containing data:
samples <- list.files(path = "./salmon_output", full.names = T) # change the path to the actual location of SalmonOutput folder
samples

# Obtain a vector of all filenames including the path:
files <- file.path(samples, "quant.sf")
files

# Name elements of files object:
names(files) <- str_replace(samples, "./salmon_output", "")
files

# Run tximport (important arguments see below):
# ignoreTxVersion = split the tx id on the '.' character to remove
# ignoreAfterBar = to split the tx id on the '|' character
txi <- tximport(files, 
                type="salmon", 
                tx2gene=tx2gene[,c("tx_id", "gene_id")],
                countsFromAbundance="lengthScaledTPM",
                ignoreAfterBar = TRUE, ignoreTxVersion = TRUE) 

# View attributes of txi
attributes(txi)

# Write the counts to an object:
data <- data.frame(round(txi$counts))

# Save the data as a table:
write.table(data, "./data/raw_data.csv", sep=';', quote=F, row.names=F)

# Create a metadata:
genotype <- factor(c(rep("experiment", 3), rep("control",3)))
meta <- data.frame(genotype, row.names = colnames(data))
meta
```

## STEP3. Normalization.

```{r}
# Create DESeq2Dataset (DDS) object:
dds <- DESeqDataSetFromMatrix(countData = round(data), colData = meta, design = ~ genotype)
dds

#  Perform normalization:
dds <- estimateSizeFactors(dds)
sizeFactors(dds)

# Take a look at the total number of reads for each sample:
colSums(counts(dds))

#  Now, to retrieve the normalized counts matrix from dds, use the counts() function and add the argument normalized=TRUE.
normalized_counts <- counts(dds, normalized=TRUE)
write.table(normalized_counts, 
            file="./data/normalized_counts.txt", #change the path to your working directory
            sep="\t", quote=F, col.names=NA)
```

## STEP4. Quality Control.

```{r}
# Perform rlog-transformation:
rld <- rlog(dds, blind=TRUE)

# Create function to plot PCA:
plotPCA.custom <- function (object, intgroup = "condition", ntop = 500, returnData = FALSE)
{
  font.size <- 15
  rv <- rowVars(assay(object))
  r <- assay(object)
  select <- order(rv, decreasing = TRUE)[seq_len(min(ntop, length(rv)))]
  pca <- prcomp(t(assay(object)[select, ]))
  percentVar <- pca$sdev^2/sum(pca$sdev^2)
  
  intgroup.df <- as.data.frame(colData(object)[, intgroup, drop = FALSE])
  group <- if (length(intgroup) > 1) {
    factor(apply(intgroup.df, 1, paste, collapse = " : "))
  }
  else {
    colData(object)[[intgroup]]
  }
  d1 <- data.frame(PC1 = pca$x[, 1], PC2 = pca$x[, 2], intgroup.df, name = colnames(object))
  
  p1 <- ggplot(data = d1, aes_string(x = "PC1", y = "PC2")) + 
    geom_point(aes_string(color = "genotype"), size = 3) + 
    xlab(paste0("PC1: ", round(percentVar[1] * 100), "% variance")) + 
    ylab(paste0("PC2: ", round(percentVar[2] * 100), "% variance")) +
    theme_bw()
  
  return (p1)
}
# Plot PCA graph:
plotPCA.custom(rld, intgroup="genotype")

ggsave('./plots/PCAplot.pdf', height = 7, width = 7)

# Correlation Heatmap
#Extract the rlog matrix from the object:
rld_mat <- assay(rld)

# Compute pairwise correlation values
rld_cor <- cor(rld_mat) 

# Plot the correlation values as a heatmap:
pheatmap(rld_cor, annotation = meta)
pheatmap(rld_cor, annotation = meta, filename = './data/samples_heatmap.pdf', width = 7, height = 5)
```

## STEP5. Statistical Testing.

```{r}
dds_analysis <- DESeq(dds)

#Plot dispertion estimates:
png("dispersion.png", width = 3000, height = 3000, res = 500)
plotDispEsts(dds_analysis)
dev.off()

#Specify contrasts to indicate the two groups we want to compare:
contrast <- c("genotype", "experiment", "control")

# First, calculate unshrunken results:
res_unshrunken <- results(dds_analysis, contrast=contrast, alpha = 0.05)

# Apply fold change shrinkage:
res <- lfcShrink(dds_analysis, contrast=contrast, res=res_unshrunken, type='normal')
write.table(res, "./data/results.csv", sep=';', quote=F, row.names=F)

#Plot the shrunken results:
plotMA(res, ylim=c(-3,3))

pdf("MA_plot_shrunken.pdf")
plotMA(res, ylim=c(-3,3))
dev.off()

# Check the results summary:
summary(res, alpha = 0.05)

#Set thresholds:
padj.cutoff <- 0.05
lfc.cutoff <- 0.58

# Convert the results table into a tibble:
res_tb <- res %>% data.frame() %>% rownames_to_column(var="gene") %>% as_tibble()

# Select significant genes only:
sig <- res_tb %>%dplyr::filter(padj < padj.cutoff & abs(log2FoldChange) >= lfc.cutoff)

#Save as a .csv table:
write.table(sig, "./data/significant_genes.csv", sep=';', quote=F, row.names=F)
```

## STEP6. Visualizing the DE analysis results.

```{r}
#Volcano plot
# Create a data frame from res object:
res.df <- as.data.frame(res)
head(res.df)

# First, create a column with gene_ids:
res.df$gene_id <- rownames(res.df)

#Join tables:
res.df.an <- left_join(res.df, genedb2, by='gene_id')

EnhancedVolcano(res.df.an,
                lab = res.df.an$symbol, 
                x = 'log2FoldChange',
                y = 'padj', 
                selectLab = NA,
                title = 'experiment vs control',
                pCutoff = 0.05, 
                FCcutoff = 0.58,
                boxedLabels = TRUE)

ggsave('./plots/Volcano_plot.pdf', height = 7, width = 7)

# Convert normalized_counts to a data frame and 
# transfer the row names to a new column called “gene”
normalized_counts <- counts(dds_analysis, normalized=T) %>%
  data.frame() %>%
  rownames_to_column(var="gene")

# Extract normalized expression for significant genes:
norm_sig <- normalized_counts %>%
  filter(gene %in% sig$gene)

# Create a heatmap for all genes:
pheatmap(norm_sig[2:7],
         cluster_rows = T,
         show_rownames = F,
         annotation = meta,
         border_color = NA,
         fontsize = 10,
         scale = "row",
         fontsize_row = 10,
         height = 10)

ggsave('./plots/genes_heatmap.pdf', height = 7, width = 7)
```

## STEP7. Functional Analysis.

```{r}
#Add the symbol column to sig object:
sig_an <- left_join(sig, genedb2, by=c('gene'='gene_id'))

#Split the data to the up- and down-regulated genes
sig_an_up <- sig_an %>% filter(log2FoldChange > 0.58)

write.table(sig_an_up, "./data/UPreg_genes.csv", sep=';', quote=F, row.names=T)

sig_an_down <- sig_an %>% filter(log2FoldChange < -0.58)

write.table(sig_an_down, "./data/DOWNreg_genes.csv", sep=';', quote=F, row.names=T)

# Run Gene Ontology (GO) analysis:
# BP - biological processes, MF - molecular functions, 
# CC - cellular components

ego_BP_up <- enrichGO(gene = sig_an_up$gene,
                      universe = res_tb$gene, 
                      keyType = "ENSEMBL",
                      OrgDb = org.Hs.eg.db, 
                      ont = "BP",
                      pAdjustMethod = "BH", 
                      pvalueCutoff = 0.05)

# Create a data frame from ego object:
d.ego_BP_up <- data.frame(ego_BP_up)

options(enrichplot.colours = c("red","blue"))
barplot(ego_BP_up, showCategory=10)
dotplot(ego_BP_up, showCategory = 10)


ego_BP_down <- enrichGO(gene = sig_an_down$gene,
                        universe = res_tb$gene, 
                        keyType = "ENSEMBL",
                        OrgDb = org.Hs.eg.db, 
                        ont = "BP",
                        pAdjustMethod = "BH", 
                        pvalueCutoff = 0.05)

# Create a data frame from ego object:
d.ego_BP_down <- data.frame(ego_BP_down)

barplot(ego_BP_down, showCategory=10)
dotplot(ego_BP_down, showCategory = 10)

# Create a data frame from ego_MF object:
ego_MF_up <- enrichGO(gene = sig_an_up$gene,
                      universe = res_tb$gene, 
                      keyType = "ENSEMBL",
                      OrgDb = org.Hs.eg.db, 
                      ont = "MF",
                      pAdjustMethod = "BH", 
                      pvalueCutoff = 0.05)

d.ego_MF_up <- data.frame(ego_MF_up)

barplot(ego_MF_up, showCategory=10)
dotplot(ego_MF_up, showCategory = 10)

ego_MF_down <- enrichGO(gene = sig_an_down$gene,
                        universe = res_tb$gene, 
                        keyType = "ENSEMBL",
                        OrgDb = org.Hs.eg.db, 
                        ont = "MF",
                        pAdjustMethod = "BH", 
                        pvalueCutoff = 0.05)

d.ego_MF_down <- data.frame(ego_MF_down)

barplot(ego_MF_down, showCategory=10)
dotplot(ego_MF_down, showCategory = 10)

# Create a data frame from ego_CC object:
ego_CC_up <- enrichGO(gene = sig_an_up$gene,
                      universe = res_tb$gene, 
                      keyType = "ENSEMBL",
                      OrgDb = org.Hs.eg.db, 
                      ont = "CC",
                      pAdjustMethod = "BH", 
                      pvalueCutoff = 0.05)

d.ego_CC_up <- data.frame(ego_CC_up)

barplot(ego_CC_up, showCategory=10)
dotplot(ego_CC_up, showCategory = 10)

ego_CC_down <- enrichGO(gene = sig_an_down$gene,
                        universe = res_tb$gene, 
                        keyType = "ENSEMBL",
                        OrgDb = org.Hs.eg.db, 
                        ont = "CC",
                        pAdjustMethod = "BH", 
                        pvalueCutoff = 0.05)

d.ego_CC_down <- data.frame(ego_CC_down)

barplot(ego_CC_down, showCategory=10)
dotplot(ego_CC_down, showCategory = 10)
```

```{r include=FALSE}
# save the dotplots
pdf("./plots/dotplot_BP_UP.pdf", width = 7, height = 7)
dotplot(ego_BP_up, showCategory = 10)
dev.off()

pdf("./plots/dotplot_BP_DOWN.pdf", width = 7, height = 7)
dotplot(ego_BP_down, showCategory = 10)
dev.off()

pdf("./plots/dotplot_BP_UP.pdf", width = 7, height = 7)
dotplot(ego_MF_up, showCategory = 10)
dev.off()

pdf("./plots/dotplot_BP_DOWN.pdf", width = 7, height = 7)
dotplot(ego_MF_down, showCategory = 10)
dev.off()

pdf("./plots/dotplot_BP_UP.pdf", width = 7, height = 7)
dotplot(ego_CC_up, showCategory = 10)
dev.off()

pdf("./plots/dotplot_BP_DOWN.pdf", width = 7, height = 7)
dotplot(ego_CC_down, showCategory = 10)
dev.off()
```

```{r}
# Plotting the plot with all 3 (BP, MF, CC) categories:

# Create a column "Category" for each table of up- and down-regulated genes:
d.ego_BP_up <- mutate(d.ego_BP_up, Category = "Biological processes")
d.ego_MF_up <- mutate(d.ego_MF_up, Category = "Molecular functions")
d.ego_CC_up <- mutate(d.ego_CC_up, Category = "Cellular components")

d.ego_BP_down <- mutate(d.ego_BP_down, Category = "Biological processes")
d.ego_MF_down <- mutate(d.ego_MF_down, Category = "Molecular functions")
d.ego_CC_down <- mutate(d.ego_CC_down, Category = "Cellular components")

# Choose the first 10 rows of each table:
d.ego_BP_up_10 <- head(d.ego_BP_up, 10)
d.ego_MF_up_10 <- head(d.ego_MF_up, 10)
d.ego_CC_up_10 <- head(d.ego_CC_up, 10)

d.ego_BP_down_10 <- head(d.ego_BP_down, 10)
d.ego_MF_down_10 <- head(d.ego_MF_down, 10)
d.ego_CC_down_10 <- head(d.ego_CC_down, 10)

# Merge these tables to one table:
d.categories_up <- bind_rows(d.ego_BP_up_10, d.ego_MF_up_10, d.ego_CC_up_10)
head(d.categories_up)

d.categories_down <- bind_rows(d.ego_BP_down_10, d.ego_MF_down_10, d.ego_CC_down_10)
head(d.categories_down)

# Creating the common plot
result_u <- d.categories_up %>%
  arrange(Category, desc(Count)) 

result_u <- result_u %>% arrange(desc(row_number()))

result_u$Description <- factor(result_u$Description, levels = result_u$Description)

gg_up <- ggplot(result_u, aes(x = Count, y = Description, fill = Category)) +
  geom_bar(stat = "identity", position = "dodge") +
  ggtitle("Gene Ontology (GO) term enrichment analysis for up-regulated genes") +
  theme_minimal() +
  scale_fill_manual(values = c("Biological processes" = "steelblue", "Molecular functions" = "purple", "Cellular components" = "orange"))

ggsave("upregulated_genes_go_enrichment.png", plot = gg_up, width = 10, height = 6, units = "in", dpi = 300)

# Creating the common plot
result_d <- d.categories_down %>%
  arrange(Category, desc(Count)) 

result_d <- result_d %>% arrange(desc(row_number()))

result_d$Description <- factor(result_d$Description, levels = result_d$Description)

gg_down <- ggplot(result_d, aes(x = Count, y = Description, fill = Category)) +
  geom_bar(stat = "identity", position = "dodge") +
  ggtitle("Gene Ontology (GO) term enrichment analysis for down-regulated genes") +
  theme_minimal() +
  scale_fill_manual(values = c("Biological processes" = "steelblue", "Molecular functions" = "purple", "Cellular components" = "orange"))

ggsave("downregulated_genes_go_enrichment.png", plot = gg_down, width = 10, height = 6, units = "in", dpi = 300)
```

## STEP8. KEGG Patway Analysis

```{r}
# KEGG analysis
entrez = bitr(sig_an$gene, 
              fromType="ENSEMBL", 
              toType="ENTREZID",
              OrgDb="org.Hs.eg.db")

sig_entrez <- left_join(sig_an, entrez, by=c('gene'='ENSEMBL'))
head(sig_entrez)

ekegg <- enrichKEGG(gene = sig_entrez$ENTREZID,
                    organism = 'human',
                    pvalueCutoff = 0.05)


fc <- sig_entrez$log2FoldChange
names(fc) <- sig_entrez$ENTREZID
fc.sorted <- sort(fc, decreasing = TRUE) 

pathview(gene.data = fc.sorted,
         pathway.id = "hsa04360",
         species = "human")
```

## STEP 9. Integration of transcriptomic and proteomic data.

```{r include=FALSE}
#the protein table you upload should contain LFC, gene symbol, UNIPROTID and padj values!
protein_table <- read.csv("./knockout122_with_symbols_padj.csv", header = TRUE, sep = ";")

#rename column 13 'Abundance Ratio (log2)' to LFC_prot 
colnames(protein_table)[13] <- "LFC_prot"

# select the necessary columns
protein_table <- protein_table  %>% select("LFC_prot", 
                               "SYMBOL", "UNIPROTID", 
                               "adjusted_p_value")


# upload the transcriptomic data
res_tb <- left_join(res.df, genedb2, by='gene_id')

#rename 'symbol' to 'SYMBOL'
res_tb <- res_tb %>% rename(SYMBOL = symbol)


# merge the transcriptomic 'res_tb' and proteomic 'protein_table'  
merged <- left_join(protein_table, res_tb, by='SYMBOL')
merged %>% filter(LFC_prot < 0 & log2FoldChange > 0)  %>% view() 

write.csv(merged, file="prot_rna_corr.csv", sep=';', quote=F, row.names=F)
```

```{r message=TRUE}
cor_pearson <- cor(merged$LFC_prot, merged$log2FoldChange, method = "pearson", use = "complete.obs")
print(cor_pearson)

cor_spearman <- cor(merged$LFC_prot, merged$log2FoldChange, method = "spearman", use = "complete.obs")
print(cor_spearman)

#change the list of genes of your interest
highlight_genes <- c("ARHGEF10L", "ARHGEF5", "ARHGEF9",
                     "CDK5", "CDK5R1", "CDK5R2", 
                     "EFNB3", "EPHA2", "EPHA5", "EPHA6", "EPHB1", "EPHB3", 
                     "MYH14", "MYH7", "MYH9","MYL12A","MYL12B", "MYL6", "MYL9",
                     "PAK3", "PAK6", "PLXNA1", "PLXNA3", "PLXNB3", 
                     "SEMA4B", "SEMA4G", "SEMA6C", "LIMK1")

red_points <- merged %>% filter(SYMBOL %in% highlight_genes)

png("./plots/correlation_plot.png", width = 3500, height = 3500, res = 500)

ggplot(merged, aes(x = log2FoldChange, y = LFC_prot)) +
  geom_point(alpha = 0.5, size = 1.5, color = "gray") +
  geom_smooth(method = "lm", se = FALSE, color = "darkgray", show.legend = FALSE) +
  geom_point(data = red_points, aes(color = SYMBOL), size = 3, show.legend = FALSE) + 
  geom_point(data = merged %>% filter(SYMBOL %in% highlight_genes),
             aes(color = SYMBOL), size = 3) +
  ylim(-4, 4) + xlim(-4, 4) +
  labs(x = "logFoldChange RNA", y = "logFoldChange protein") +
  scale_color_manual(values = rep("red", length(highlight_genes))) +
  theme_bw() +
  geom_label_repel(data = merged %>% filter(SYMBOL %in% highlight_genes), aes(label = SYMBOL), color = "black", size = 3, label.padding = unit(0.2, "lines"), box.padding = unit(0.4, "lines"), label.background = ggplot2::element_rect(fill = "white", colour = NA)) +
  theme(legend.position = "none",
        axis.line = element_line(color = "black")) + 
  geom_vline(xintercept = 0, linetype = "dashed", color = "black") + 
  geom_hline(yintercept = 0, linetype = "dashed", color = "black") 

dev.off()
```